0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,262 @@
+5. The sender quotes and sends their payment, setting the data field of the ILP packet to the PSK data.
+6. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
+7. If the public `Encryption` header is set, then the receiver derives the payment encryption key from the pre-shared secret key, sets the AES-256-CTR initialization vector to the `Nonce` header in the PSK public headers, and uses them to decrypt the private headers and their data. If the public `Encryption` header is set to `none`, the receiver parses the private headers and their data in clear-text.
+8. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call an external system to make sure the incoming funds are expected.
+9. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the same HMAC key as the sender, derived from the shared secret.
+
+## Data Format
+
+The PSK details are divided into two sets of headers. Each set of headers is formatted like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Application data is appended to the private headers, after a blank line. This object is then encrypted and appended to the public headers as binary data, after a blank line. Both private and public headers are parsed in the exact same way as HTTP headers.
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Nonce: fpwpAhlN588
+Key: hmac-sha-256
+Encryption: aes-256-ctr
+Header: data_everyone_can_see
+
+...
+```
+
+The public headers come after a single status line, which must be `PSK/1.0`. Public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+PSK payments use an HMAC of the PSK details as a fulfillment, to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The decryption key is derived from the pre-shared secret key, and the AES-256-CTR initialization vector is set to the value of the `Nonce` header (`fpwpAhlN588` in the above example). The nonce MUST be generated with cryptographically-secure randomness. **If the nonce is reused with the same shared secret, it could leak unencrypted data or allow money to be stolen by malicious parties.**
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ The HMAC algorithm, which MUST be `hmac-sha-256` (case sensitive) |
```

Comment:
This is not really the HMAC algorithm, it's the algorithm for generating the key to be used for the encryption and fulfillment generation. If you use ECDH to generate that value it would be something like `ecdh-secp256k1 <public key>`.

I don't think we should say this MUST  be `hmac-sha-256`, I think we should have a list of supported algorithms and say that for PSK 1.0 implementations must at minimum support this algorithm. When it says "MUST", it makes me think that you'd automatically reject anything that doesn't say that and any other value would technically not conform to the spec.